src/tallies/tally_scoring.cpp

```diff
@@ -0,0 +1,2347 @@
+#include "openmc/tallies/tally_scoring.h"
+
+#include "openmc/bank.h"
+#include "openmc/capi.h"
+#include "openmc/constants.h"
+#include "openmc/error.h"
+#include "openmc/material.h"
+#include "openmc/mgxs_interface.h"
+#include "openmc/nuclide.h"
+#include "openmc/photon.h"
+#include "openmc/reaction_product.h"
+#include "openmc/search.h"
+#include "openmc/settings.h"
+#include "openmc/simulation.h"
+#include "openmc/string_utils.h"
+#include "openmc/tallies/derivative.h"
+#include "openmc/tallies/filter.h"
+#include "openmc/tallies/filter_delayedgroup.h"
+#include "openmc/tallies/filter_energy.h"
+
+#include <string>
+
+namespace openmc {
+
+//==============================================================================
+// FilterBinIter implementation
+//==============================================================================
+
+FilterBinIter::FilterBinIter(const Tally& tally, const Particle* p)
+  : tally_{tally}
+{
+  // Find all valid bins in each relevant filter if they have not already been
+  // found for this event.
+  for (auto i_filt : tally_.filters()) {
+    auto& match {simulation::filter_matches[i_filt]};
+    if (!match.bins_present_) {
+      match.bins_.clear();
+      match.weights_.clear();
+      model::tally_filters[i_filt]->get_all_bins(p, tally_.estimator_, match);
+      match.bins_present_ = true;
+    }
+
+    // If there are no valid bins for this filter, then there are no valid
+    // filter bin combinations so all iterators are end iterators.
+    if (match.bins_.size() == 0) {
+      index_ = -1;
+      return;
+    }
+
+    // Set the index of the bin used in the first filter combination
+    match.i_bin_ = 0;
+  }
+
+  // Compute the initial index and weight.
+  this->compute_index_weight();
+}
+
+FilterBinIter::FilterBinIter(const Tally& tally, bool end)
+  : tally_{tally}
+{
+  // Handle the special case for an iterator that points to the end.
+  if (end) {
+    index_ = -1;
+    return;
+  }
+
+  for (auto i_filt : tally_.filters()) {
+    auto& match {simulation::filter_matches[i_filt]};
+    if (!match.bins_present_) {
+      match.bins_.clear();
+      match.weights_.clear();
+      for (auto i = 0; i < model::tally_filters[i_filt]->n_bins(); ++i) {
+        match.bins_.push_back(i);
+        match.weights_.push_back(1.0);
+      }
+      match.bins_present_ = true;
+    }
+
+    if (match.bins_.size() == 0) {
+      index_ = -1;
+      return;
+    }
+
+    match.i_bin_ = 0;
+  }
+
+  // Compute the initial index and weight.
+  this->compute_index_weight();
+}
+
+FilterBinIter&
+FilterBinIter::operator++()
+{
+  // Find the next valid combination of filter bins.  To do this, we search
+  // backwards through the filters until we find the first filter whose bins
+  // can be incremented.
+  bool done_looping = true;
```

Comment:
Optional: As an alternative to using the `done_looping` flag, your could declare the `i` index outside the for loop and check that variable in the `if` block below.  This would decrease the amount of state you have to track.  Renaming it would make it more descriptive as well.  E.g.:

```
int tally_index = tally_.filters().size() - 1;
for (; tally_index >= 0; tally_index--) {
   ...
}

// Do this instead of checking `done_looping`
if (tally_index == 0) {
   ...
}
```

If you choose not to do the above, would also recommend choosing a variable name more tied to the business logic than the procedural logic.  Maybe instead of `done_looping`, use `checked_all_tallies` or `visited_every_combination`.